Feat: Weekly view UI

eleonoraka opened the pull request:
## Summary

Added new feature weekly view screen. Now trips can be seen weekly. When clicking on a day of the week, takes to by day screen.

## Changes

- **Screens/UI:** WeeklyViewScreen
- **Bug Fixes/Improvements:** closes issue #87 

## Checklist

- [x] This PR resolves #87 
- [x] Tests have been added for new functionality.

![WhatsApp Image 2024-11-09 at 4 10 55 PM](https://github.com/user-attachments/assets/bcadaabe-c9d7-4712-99dd-5b26caca51d5)


##  Testing

- **Manual Testing:**
    - [x] Tested on emulator / physical device.

##  Related Issues/Tickets

Closes #87

BogdanCDev commented on app/src/main/java/com/android/voyageur/ui/trip/schedule/ScheduleScreen.kt:
```diff
@@ -0,0 +1,63 @@
+                fontWeight = if (isDailySelected) FontWeight.Bold else FontWeight.Normal)
+          }
+          Text(text = " / ", style = MaterialTheme.typography.bodyLarge, color = Color.Gray)
+          TextButton(onClick = { isDailySelected = false }) {
+            Text(
+                text = "Weekly",
+                style = MaterialTheme.typography.bodyLarge,
+                color = if (!isDailySelected) Color.Black else Color.Gray,
+                fontWeight = if (!isDailySelected) FontWeight.Bold else FontWeight.Normal)
+          }
```
You can add content descriptions to increase accesibility

BogdanCDev commented on app/src/main/java/com/android/voyageur/ui/trip/schedule/ScheduleScreen.kt:
```diff
@@ -0,0 +1,63 @@
+        modifier = Modifier.fillMaxWidth().padding(bottom = 8.dp, end = 16.dp),
+        horizontalArrangement = Arrangement.End,
+        verticalAlignment = Alignment.CenterVertically) {
+          TextButton(onClick = { isDailySelected = true }) {
+            Text(
+                text = "Daily",
+                style = MaterialTheme.typography.bodyLarge,
+                color = if (isDailySelected) Color.Black else Color.Gray,
+                fontWeight = if (isDailySelected) FontWeight.Bold else FontWeight.Normal)
+          }
```
You can add content descriptions to increase accesibility

BogdanCDev commented on app/src/main/java/com/android/voyageur/ui/trip/schedule/WeeklyView.kt:
```diff
@@ -0,0 +1,179 @@
+    onDaySelected: () -> Unit // Add this parameter
+) {
+  Card(
+      modifier = Modifier.width(360.dp).testTag("weekCard_$weekIndex"),
```
Might not adapt well to all screen sizes due to the hard coded size, you can use something like fillMaxWidth

eleonoraka replied:
Thank you for the remark! I agree with you, I didn't think of it at first. I will fix it in the next PR.

BogdanCDev commented on app/src/main/java/com/android/voyageur/ui/trip/schedule/WeeklyView.kt:
```diff
@@ -0,0 +1,179 @@
+}
+
+fun formatDate(date: LocalDate): String {
+  return "${date.month.toString().take(3)} ${date.dayOfMonth}"
```
might be better to use DateTimeFormatter, because it might be useful to take into account also the user's locale settings

eleonoraka replied:
Thank you for the remark! I agree with you, I changed to DateTimeFormatter.

BogdanCDev commented on app/src/main/java/com/android/voyageur/ui/trip/schedule/WeeklyView.kt:
```diff
@@ -0,0 +1,179 @@
+      }
+}
+
+fun generateWeeks(startTimestamp: Timestamp, endTimestamp: Timestamp): List<DateRange> {
+  val startDate = startTimestamp.toDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
+  val endDate = endTimestamp.toDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
+
+  val weeks = mutableListOf<DateRange>()
+
+  var currentWeekStart = startDate.with(DayOfWeek.MONDAY)
+  val lastWeekStart = endDate.with(DayOfWeek.MONDAY)
+
+  while (currentWeekStart <= lastWeekStart) {
+    weeks.add(DateRange(first = currentWeekStart, last = currentWeekStart.plusDays(6)))
+    currentWeekStart = currentWeekStart.plusWeeks(1)
+  }
+
+  return weeks
+}
```
It works fine, but it also includes some dates before the start of the trip if the trip does not start on a Monday. It might be worth taking a look as a possible improvement.

ioanajianu replied:
I agree, maybe the big card can remain in this case Nov 4 - Nov 10, but inside the card we could have only the days which are part of the trip

eleonoraka replied:
Thank you for your remarks! I agree with you, I changed the code and fixed this issue.

ioanajianu commented:
Can you add some screenshots/screen recordings to the PR description please?
Also, maybe try to make the UI a bit more similar to the Figma mockup and the other screens.
Right now the text and card sizes are significantly bigger than what's in the mockup. I like the idea of the daily cards being not too small, so that you can click on them to direct you to the daily view, but I still think they should be a bit smaller

mirceagavrizi commented on app/src/androidTest/java/com/android/voyageur/ui/trip/schedule/WeeklyViewScreenTest.kt:
```diff
@@ -0,0 +1,196 @@
+                                    LocalDateTime.of(2024, 10, 3, 14, 0)
+                                        .toInstant(ZoneOffset.UTC))),
+                        endDate =
+                            Timestamp(
```
Here, endDate is now endTime, can you adjust the changes?

eleonoraka replied:
Thank you for making me notice these changes! I fixed this in the code.

ioanajianu commented:
Can you please add the Plus button on this screen to open the add activity screen? It would be more user friendly to be able to add activities regardless of Daily/Weekly view

ioanajianu commented:
I have a question about the intended behavior when clicking on a day of the week: I see it redirects to the ByDay screen. Did you intend for it to also scroll to the particular day you press on?
This would be a nice feature to have, of course not mandatory for now.
However from a user experience point of view, it's a bit weird to click on a day that's a few weeks from now and get redirected to the beginning of the trip on the daily view. 
If you can't implement the scrolling to the particular day you clicked on from the Daily view, I can redirect to the screen that shows the activities for one day which I will implement #122

eleonoraka commented:
> I have a question about the intended behavior when clicking on a day of the week: I see it redirects to the ByDay screen. Did you intend for it to also scroll to the particular day you press on? This would be a nice feature to have, of course not mandatory for now. However from a user experience point of view, it's a bit weird to click on a day that's a few weeks from now and get redirected to the beginning of the trip on the daily view. If you can't implement the scrolling to the particular day you clicked on from the Daily view, I can redirect to the screen that shows the activities for one day which I will implement #122

Thanks for asking about the day selection behavior! 
You're right - clicking a day currently just navigates to ByDayScreen. I actually think your suggestion to redirect to a single day view would be the better UX approach here. It's a very nice idea! :)

eleonoraka commented:
> Can you please add the Plus button on this screen to open the add activity screen? It would be more user friendly to be able to add activities regardless of Daily/Weekly view

I fixed the code and added the button which opens the add activity screen!

ioanajianu reviewed (approved):
Thanks for implementing the changes! LGTM
